### PR TITLE
Support for stateful policies

### PR DESCRIPTION
**Pull request by timbauman**
## Description

Add support for stateful policies

https://github.com/HumanCompatibleAI/imitation/issues/413

## Testing

Unit tests pass, plus new unit test.

The notebook at docs/tutorials/6_train_mce.ipynb learns a slightly different reward function - the expert policies are found with the soft bellman backup, which has a more variable behavior in later time steps. @levmckinney said this is likely due to the fact that there's less future reward so the algorithm focuses more on maximizing entropy. Over longer time horizons he said the old and new policies should be more similar, so I set the horizon to 40. Both learned reward functions changed, but they are more similar (and the new policy is less variable)

Without state (horizon 8):
image

With state (horizon 8):

ex 1
image

ex 2
image

ex 3
image

Without state (horizon 40):
image

With state (horizon 40):

image

### CODE REVIEW

File: src/imitation/policies/exploration_wrapper.py [hunk 1]
```diff
@@ -53,9 +55,14 @@ def __init__(
...
-    def _random_policy(self, obs: np.ndarray) -> np.ndarray:
+    def _random_policy(
+        self,
+        obs: np.ndarray,
+        state: Optional[Tuple[np.ndarray, ...]],
+        episode_start: Optional[np.ndarray],
+    ) -> Tuple[np.ndarray, Optional[Tuple[np.ndarray, ...]]]:
         acts = [self.venv.action_space.sample() for _ in range(len(obs))]
-        return np.stack(acts, axis=0)
+        return np.stack(acts, axis=0), None  # is this right?
```
Comment (timbauman) [inline src/imitation/policies/exploration_wrapper.py:65]: This feels like a weird case. If you have a stateful policy, it'll randomly lose all state every time exploration happens. On the other hand, if you just pass the state through this policy, it'll behave strangely since the world state will have changed without the hidden state having a chance to catch up. I think None is an okay solution - or, we could throw or warn if the state passed to the random policy is not None to indicate that the behavior is undefined?

Comment (timbauman) [reply src/imitation/policies/exploration_wrapper.py:65]: Or we could simply always pass an empty state to the wrapped policy

Comment (levmckinney) [reply src/imitation/policies/exploration_wrapper.py:65]: Stateful polices and exploration don't seem like they go together very well. I'm struggling to think of a single example of an off policy stateful algorithm that is used in practice. This should definitely have some kind of warning, maybe even an error. I would bump this one up to @AdamGleave.

Comment (timbauman) [reply src/imitation/policies/exploration_wrapper.py]: Updated to throw. (Talked with @AdamGleave offline and he agrees)

File: src/imitation/data/rollout.py [hunk 2]
```diff
@@ -274,26 +278,38 @@ def policy_to_callable(
...
-        def get_actions(states):
-            acts = [venv.action_space.sample() for _ in range(len(states))]
-            return np.stack(acts, axis=0)
+        def get_actions(
+            observation: np.ndarray,
+            state: Optional[Tuple[np.ndarray, ...]],
+            episode_start: Optional[np.ndarray],
+        ) -> Tuple[np.ndarray, Optional[Tuple[np.ndarray, ...]]]:
+            acts = [venv.action_space.sample() for _ in range(len(observation))]
+            return np.stack(acts, axis=0), None  # is None the right state?
```
Comment (levmckinney) [inline src/imitation/data/rollout.py:290]: Yes the random agent isn't stateful.

File: src/imitation/data/rollout.py [hunk 3]
```diff
@@ -274,26 +278,38 @@ def policy_to_callable(
...
 
     elif isinstance(policy, (BaseAlgorithm, BasePolicy)):
         # There's an important subtlety here: BaseAlgorithm and BasePolicy
         # are themselves Callable (which we check next). But in their case,
         # we want to use the .predict() method, rather than __call__()
         # (which would call .forward()). So this elif clause must come first!
 
-        def get_actions(states):
+        def get_actions(
```
Comment (levmckinney) [inline src/imitation/data/rollout.py:298]: nitpick: this should be `observations`, `states` and `episode_starts` if we want to be consistent.

File: src/imitation/policies/exploration_wrapper.py [hunk 4]
```diff
@@ -64,8 +71,20 @@ def _switch(self) -> None:
...
+        observation: np.ndarray,
+        state: Optional[Tuple[np.ndarray, ...]],
+        _episode_start: Optional[np.ndarray],
+    ) -> Tuple[np.ndarray, Optional[Tuple[np.ndarray, ...]]]:
+        if state is not None:
+            raise ValueError("Exploration wrapper does not support stateful policies.")
+
+        acts, state = self.current_policy(observation, None, None)
+
+        if state is not None:
```
Comment (levmckinney) [inline src/imitation/policies/exploration_wrapper.py:85]: Do we need both of these checks?

**Review by levmckinney — APPROVED** [review]
Some very minor nitpicks. Overall this looks good.

File: src/imitation/policies/exploration_wrapper.py (hunk 4 above)
Comment (timbauman) [reply src/imitation/policies/exploration_wrapper.py:85]: I guess technically you could only use one, but they are checking different things. Probably the check for the return value is more critical. But I'll add comments explaining the rationale for having both for now

File: src/imitation/data/rollout.py [hunk 5]
```diff
@@ -264,7 +265,10 @@ def sample_until(trajs: Sequence[types.TrajectoryWithRew]) -> bool:
 
 # A PolicyCallable is a function that takes an array of observations
```
Comment (AdamGleave) [inline src/imitation/data/rollout.py:266]: Comment is inconsistent with the new definition

File: src/imitation/data/rollout.py [hunk 6]
```diff
@@ -274,26 +278,38 @@ def policy_to_callable(
     deterministic_policy: bool = False,
 ) -> PolicyCallable:
     """Converts any policy-like object into a function from observations to actions."""
+    get_actions: PolicyCallable
```
Comment (AdamGleave) [inline src/imitation/data/rollout.py:281]: :+1:

File: src/imitation/policies/exploration_wrapper.py [hunk 7]
```diff
@@ -53,9 +55,14 @@ def __init__(
         # Choose the initial policy at random
         self._switch()
 
-    def _random_policy(self, obs: np.ndarray) -> np.ndarray:
+    def _random_policy(
+        self,
+        obs: np.ndarray,
+        _state: Optional[Tuple[np.ndarray, ...]],
+        _episode_start: Optional[np.ndarray],
+    ) -> Tuple[np.ndarray, Optional[Tuple[np.ndarray, ...]]]:
```
Comment (AdamGleave) [inline src/imitation/policies/exploration_wrapper.py:63]: Nit: our style guide https://google.github.io/styleguide/pyguide.html#214-decision favors deleting variables to suppress unused variable warnings.

File: src/imitation/policies/exploration_wrapper.py (hunk 4 above)
Comment (AdamGleave) [reply src/imitation/policies/exploration_wrapper.py:85]: It definitely *looks* redundant but checking both input and return value will catch errors faster and more localized (although I struggle to think of cases where an error wouldn't be caught on the next step even if you omitted one check). Adding comments to justify this sounds good. You could also consider modifying the error message a bit between them to emphasize where the non-None state is coming from.

File: tests/algorithms/test_mce_irl.py [hunk 8]
```diff
@@ -314,6 +314,31 @@ def test_tabular_policy(rng):
     np.testing.assert_equal(timesteps[0], 2 - mask.astype(int))
 
 
+def test_tabular_policy_rollouts(rng):
+    """Tests tabular policy prediction, especially timestep calculation and masking."""
```
Comment (AdamGleave) [inline tests/algorithms/test_mce_irl.py:318]: What the test is doing is a bit hard to follow, consider expanding docstring, adding comments or otherwise clarifying?

My understanding (but it took a few minutes for me to work out) is:
  - `s_a` always takes the same action (deterministically) regardless of state, and alternates round-robin between the three actions.
  - `pi` is `s_a` tiled 7 times to get up to 21 timesteps, greater than the horizon of `20` of `ReasonablePOMDP`.
  - We then check that the actions taken in fact do alternate round robin as we'd expect from the construction.

File: tests/algorithms/test_mce_irl.py [hunk 9]
```diff
@@ -314,6 +314,31 @@ def test_tabular_policy(rng):
...
+    state_space = gym.spaces.Discrete(5)
+    action_space = gym.spaces.Discrete(3)
+
+    s_a = np.stack([np.eye(3)] * 5, axis=1)
```
Comment (AdamGleave) [inline tests/algorithms/test_mce_irl.py:322]: Can we make it clear where `3` and `5` come from? e.g:
```suggestion
    s_a = np.stack([np.eye(action_space.n)] * state_space.n, axis=1)
```
or define `num_states`, `num_actions`

File: tests/algorithms/test_mce_irl.py [hunk 10]
```diff
@@ -314,6 +314,31 @@ def test_tabular_policy(rng):
...
+    pi = np.concatenate([s_a] * 7, axis=0)
```
Comment (AdamGleave) [inline tests/algorithms/test_mce_irl.py:323]: I think this does the same and seems more canonical? (I don't particularly care either way though.)
```suggestion
    pi = np.repeat(s_a, 7, axis=0)
```

File: tests/algorithms/test_mce_irl.py [hunk 11]
```diff
@@ -314,6 +314,31 @@ def test_tabular_policy(rng):
...
+    mdp = ReasonablePOMDP()
+    state_env = base_envs.ExposePOMDPStateWrapper(mdp)
+    state_venv = vec_env.DummyVecEnv([lambda: state_env])
+    trajs = rollout.generate_trajectories(
+        tabular,
+        state_venv,
+        sample_until=rollout.make_min_episodes(1),
+        rng=rng,
+    )
+    assert (trajs[0].acts == pi[:, 0, :].nonzero()[1][: len(trajs[0].acts)]).all()
```
Comment (AdamGleave) [inline tests/algorithms/test_mce_irl.py:339]: I wrote this to try and understand it, including in case it helps with clarifying test although would not recommend using this verbatim:
```suggestion
    # pi[t,s,a] is the same for every state, so drop that dimension
    expected_actions_onehot = pi[:, 0, :]
    # take the index in a (drop the indexes in t)
    expected_actions_index = expected_actions_onehot.nonzero()[1]
    assert (trajs[0].acts == expected_actions_index[: len(trajs[0].acts)]).all()
```

File: tests/policies/test_exploration_wrapper.py [hunk 12]
```diff
@@ -1,14 +1,19 @@
...
 import seals  # noqa: F401
 
 from imitation.policies import exploration_wrapper
 from imitation.util import util
 
 
-def constant_policy(obs):
-    return np.zeros(len(obs), dtype=int)
+def constant_policy(obs, state, mask):
+    return np.zeros(len(obs), dtype=int), None
```
Comment (AdamGleave) [inline tests/policies/test_exploration_wrapper.py:12]:
```suggestion
    del state, mask  # Unused
    return np.zeros(len(obs), dtype=int), None
```

File: tests/policies/test_exploration_wrapper.py [hunk 13]
```diff
@@ -1,14 +1,19 @@
...
+
+
+def fake_stateful_policy(obs, state, mask):
+    return np.zeros(len(obs), dtype=int), (np.zeros(1),)
```
Comment (AdamGleave) [inline tests/policies/test_exploration_wrapper.py:16]:
```suggestion
    del state, mask  # Unused
    return np.zeros(len(obs), dtype=int), (np.zeros(1),)
```

**Review by AdamGleave — APPROVED** [review]
LGTM, some minor changes requested but no need for a re-review.

**Comment by AdamGleave** [comment]
You also have some lint errors (which were previously obscured by the whole CI pipeline being broken :smiling_face_with_tear: but that's fixed now in #706)

File: tests/algorithms/test_mce_irl.py (hunk 8 above)
Comment (timbauman) [reply tests/algorithms/test_mce_irl.py]: Updated - hopefully it's clearer now
